src/uopool/services/uopool.rs

```diff
@@ -12,68 +17,111 @@ use crate::{
     },
 };
 use async_trait::async_trait;
-use ethers::types::{Address, U256};
+use ethers::{
+    prelude::gas_oracle::ProviderOracle,
+    providers::Middleware,
+    types::{Address, U256},
+};
 use jsonrpsee::{tracing::info, types::ErrorObject};
 use parking_lot::RwLock;
-use serde_json::json;
-use std::{collections::HashMap, sync::Arc};
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+};
 use tonic::Response;
 
 pub type UoPoolError = ErrorObject<'static>;
 
-pub struct UoPoolService {
+pub struct UoPoolService<M: Middleware> {
+    pub entry_points: Arc<HashMap<MempoolId, EntryPoint<M>>>,
     pub mempools: Arc<RwLock<HashMap<MempoolId, MempoolBox<Vec<UserOperation>>>>>,
     pub reputations: Arc<RwLock<HashMap<MempoolId, ReputationBox<Vec<ReputationEntry>>>>>,
+    pub sanity_check_results: Arc<RwLock<HashMap<UserOperationHash, HashSet<SanityCheckResult>>>>,
```

Comment:
Just curious. Why is the `sanity_check_results` needed?

Reply:
Sanity check results are needed for the simulation. During the simulation, we have to allow some OPCODES only if the paymaster is staked, which is verified here in sanity checks (so there is no need to check that again in simulation). Simulation is also performed two times, once when the user operation is received and again during the bundling.

Reply:
Make perfect sense.

Reply:
After some thinkings, I think it would be better to pipeline the sanity check and validation together instead of storing the results here because it would keep consuming the memory(or you plan to remove it after the validation?). But I am cool with the current solution right now. Let's see what's better later.

Reply:
Yeah, I was in a dilemma about what would be better.

It should be removed simultaneously as the user operation is removed from the mempool. Let's keep it as is, and we can address this later.